spec/components/Web3API_Client.md

```diff
@@ -0,0 +1,157 @@
+# Web3API Client
+
+Web3API client is intended for application developers that can use the client as a single library in their (d)apps to interact with various decentralized services using Web3API's packages.
+
+Currently, Web3API client is available in the following languages:
+ - JavaScript (TypeScript)
+
+## Client interface
+The following snippet contains the interface that client should implement.
+
+```
+interface Web3APIClient {
+  redirects(): UriRedirect[];
+  query(options: QueryApiOptions): QueryApiResult;
+}
+
+type UriRedirect {
+ // Redirect from this URI or set of URIs specified by the regex
+ from: Uri | RegExp;
+ // Destination URI or plugin that will handle invocations
+ to: Uri | (() => Plugin)
+}
+
+type QueryApiOptions {
+  // Target URI that is being queried
+  uri: Uri;
+  // GraphQL query
+  query: string | QueryDocument;
+  // Optional variables that are part of the GraphQL query
+  variables?: Record<string, any>;
+}
+
+type QueryApiResult {
+ /**
+   * Query result data. The type of this value is a named map,
+   * where the key is the method's name, and value is the [[InvokeApiResult]]'s data.
+   * This is done to allow for parallel invocations within a
+   * single query document. In case of method name collisions,
+   * a postfix of `_0` will be applied, where 0 will be incremented for
+   * each occurrence. If undefined, it means something went wrong.
+ */
+  data?: [[InvokeApiResult]];
+  // errors encountered during the query
+  errors?: Error[];
+}
+```
+
+### Initialization
+Client is initialized using `Web3APIClient` class that should support receiving:
+
+#### `redirects` array
+This in an unbounded array containing `UriRedirect` objects.
+Redirects are used for resolving Web3API URIs to decentralized protocols and their nodes.
+ 
+Upon initizalition it is filled by default redirects:
+ - IPFS - required for downloading Web3API packages
+ - Ethereum - required for smart contract interactions
+ - ENS - required for resolving domain to IPFS hashes
```

Comment:
Package interface and specification for each of preloaded redirects. If I'm not mistaken those must be implemented in every client language? Or if they depend on host methods, than host methods must be specified.

Reply:
I left them out intentionally but good question. Should the default redirects be the same and implemented in every client @dOrgJelli making them part of the spec?